engines/kingdom/detection.cpp

```diff
@@ -0,0 +1,186 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "base/plugins.h"
+
+#include "common/savefile.h"
+#include "engines/advancedDetector.h"
+#include "common/file.h"
+
+#include "kingdom/kingdom.h"
+
+
+namespace Kingdom {
+const char *KingdomGame::getGameId() const { return _gameDescription->gameId; }
+Common::Platform KingdomGame::getPlatform() const { return _gameDescription->platform; }
+}
+
+static const PlainGameDescriptor kingdomGames[] = {
+	{"kingdom", "Kingdom: The Far Reaches"},
+	{0, 0}
+};
+
+namespace Kingdom {
+
+#define MAX_SAVES 99
+
+static const ADGameDescription gameDescriptions[] = {
+	// Kingdom PC DOS Demo version, provided by Strangerke
+	{
+		"kingdom",
+		0,
+		AD_ENTRY1s("KINGDEMO.EXE", "2ce478fc13086d0ddb02a2867ba307dc", 167154),
+		Common::EN_ANY,
+		Common::kPlatformDOS,
+		ADGF_UNSTABLE | ADGF_DEMO,
+		GUIO1(GUIO_NONE)
+	},
+
+	// Kingdom PC DOS version v1.26 EN, provided by Strangerke
+	{
+		"kingdom",
+		0,
+		AD_ENTRY1s("KINGDOM.EXE", "64d3e03b963396ced402f3dc958765c0", 199693),
+		Common::EN_ANY,
+		Common::kPlatformDOS,
+		ADGF_UNSTABLE,
+		GUIO1(GUIO_NONE)
+	},
+
+	AD_TABLE_END_MARKER
+};
+
+} // End of namespace Kingdom
+
+class KingdomMetaEngine : public AdvancedMetaEngine {
+public:
+	KingdomMetaEngine() : AdvancedMetaEngine(Kingdom::gameDescriptions, sizeof(ADGameDescription), kingdomGames) {
+	}
+
+	const char *getEngineId() const override {
+		return "kingdom";
+	}
+
+	virtual const char *getName() const {
+		return "Kingdom: The Far Reaches' Engine";
+	}
+
+	virtual const char *getOriginalCopyright() const {
+		return "Kingdom: The far Reaches (C) 1995 Virtual Image Productions";
+	}
+
+	virtual bool hasFeature(MetaEngineFeature f) const;
+	virtual bool createInstance(OSystem *syst, Engine **engine, const ADGameDescription *desc) const;
+	virtual int getMaximumSaveSlot() const;
+	virtual SaveStateList listSaves(const char *target) const;
+	virtual void removeSaveState(const char *target, int slot) const;
+	SaveStateDescriptor querySaveMetaInfos(const char *target, int slot) const;
+};
+
+bool KingdomMetaEngine::hasFeature(MetaEngineFeature f) const {
+	return
+	    (f == kSupportsListSaves) ||
+	    (f == kSupportsLoadingDuringStartup) ||
+	    (f == kSupportsDeleteSave) ||
+	    (f == kSavesSupportMetaInfo) ||
+	    (f == kSavesSupportThumbnail) ||
+	    (f == kSavesSupportCreationDate);
+}
+
+bool KingdomMetaEngine::createInstance(OSystem *syst, Engine **engine, const ADGameDescription *desc) const {
+	if (desc)
+		*engine = new Kingdom::KingdomGame(syst, desc);
+
+	return desc != nullptr;
+}
+
+int KingdomMetaEngine::getMaximumSaveSlot() const {
+	return MAX_SAVES;
+}
+
+SaveStateList KingdomMetaEngine::listSaves(const char *target) const {
+	Common::SaveFileManager *saveFileMan = g_system->getSavefileManager();
```

Comment:
Given the more recent improvements for save files, you might want to consider taking advantage of it, and avoid having to worry about implementing list files, savegame meta, etc.